[EDD] 2주차 Wootube 기능 구현

kangmin46 opened the pull request:
2주차 기능 구현 목록은 다음과 같습니다.
* 동영상 수정/삭제 API
* 본인 채널 페이지 마크업 구현
* 회원 CRUD API
* 로그인 관련 Interceptor

Laterality committed:
[merge] merged existing PRs(lobby, conas) (#104)

* [refactor] VO 삭제

* [refactor] delete needless code

* [refactor] fix method name

* [refactor] fix class name, plus s

* [feat] signup page markup (#75)

* [feat] add signup.html, signup.js, signup.css

* [feat] add signup button margin-top 12px

* [fix] convert utc date to local date

kangmin46 committed:
[feat] User CRUD (#107)

* [feat] add save user

* [feat] add user delete

* [feat] add user read , add user_delete_when_already_deleted test

* [refactor] convert variable name

* [refactor] change password validate message

* [refactor] code formatting

* [refactor] delete RedirectResponse.java

* [refactor] accomodate peer review

* [refactor] accomodate peer review#2

Conatuseus committed:
[feat] video update&delete (#106)

* [feat] add video update

* [feat] add video delete

* [test] refactor delete test

* [feat] add Cros allowedMethods (get, post, put, delete)

* [refactor] add constant

* [refactor] change function location

Conatuseus committed:
[feat] channel view markup (#105)

* [feat] add channel view markup

* [feat] add user info markup

* [feat] change banner

* [feat] change update icon

* [refactor] add feedback

Conatuseus committed:
[feat] add front update & delete ajax (#108)

kangmin46 committed:
[feat] service segregation (#112)

* userService 와 videoService 를 각 각 internalService와 Service 로 분리

Laterality committed:
[fix] fix issue by change JVM's default timezone (#111)

* 서버에 관계없이 기본 타임존을 UTC로 하도록 설정

kangmin46 committed:
[feat] login  (#113)

* 로그인 기능을 추가했습니다 ㅎㅎ

kangmin46 committed:
[feat] add interceptor (#114)

* SignInInterceptor, NoSignInInterceptor 를 추가했습니다 ㅎㅎㅎ
* SignInInterceptor는 로그인한 유저가 접근하면 안되는 경로를 막아주는 역할입니다.
* NoSignInInterceptor는 로그인을 하지않은 유저가 접근하면 안되는 경로를 막아주는 역할입니다.ㅎ

Laterality committed:
[feat] update db schema for production DB (#115)

- 도메인 변경에 따른 DB 스키마 파일을 변경 합니다.

Laterality committed:
[fix] fix issue #116

Laterality committed:
[refactor] simplify interceptor logic

Laterality committed:
[refactor] rename interceptors

heebong committed:
fix issues #116 (#117)

interceptor 로직을 단순화하고, 실패하는 컨트롤러 테스트를 수정했습니다.

gracefulBrown commented on backend/src/main/java/com/woowacourse/edd/domain/User.java:
```diff
@@ -0,0 +1,74 @@
+    @Column(nullable = false)
+    private LocalDateTime updateDate;
+
+    private Boolean isDeleted;
```
wrapper class를 사용한 이유가 있나요?

Conatuseus replied:
Long은 null을 넣을 수 있어서 사용했는데,
생각해보니 Boolean은 null을 넣을 필요가 없어보이네요. 수정 했습니다!

gracefulBrown commented on backend/src/main/java/com/woowacourse/edd/presentation/controller/UserController.java:
```diff
@@ -0,0 +1,48 @@
+@RequestMapping(USER_URL)
+public class UserController {
+
+    public static final String USER_URL = "/v1/users";
```
version 관리를 고려한 path 네요 👍 
만약 v2가 생겨야 한다면 지금 구조에서 어떻게 해야할까요?
혹시 구조 변경이 필요하다면 어떻게 해야할 지 팀 내에서 의논해보아도 좋을 것 같아요

지금 단계에서 굳이 할 필요는 없지만 버전관리가 되어있는 것 같아서 생각해볼 거리를 던져봅니다ㅎㅎㅎ
(참고로 `api version maintenance` 라던지 `api version maintenance in spring` 와 같은 키워드로 검색해보면 좋을 것 같아요)

gracefulBrown commented on backend/src/test/java/com/woowacourse/edd/domain/VideoTests.java:
```diff
@@ -0,0 +1,70 @@
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class VideoTests {
```
중복 테스트가 많네요
@ParameterizedTest @MethodSource @CsvSource 등을 활용하여 중복을 제거해보세요 :)

Conatuseus replied:
오!!!!!!! 브라운 덕분에 아주 좋은 테스트 방법을 알아갑니다.
너무 감사합니다 👍  💯  😄  🎂 🎉 🎉

gracefulBrown commented on backend/src/test/java/com/woowacourse/edd/presentation/controller/VideoControllerTests.java:
```diff
@@ -98,7 +96,53 @@ void save_invalid_contents() {
+
+        updateVideo(id, videoUpdateRequestDto, getDefaultLoginSessionId())
+            .isOk()
+            .expectHeader()
```
다른 expectXXX는 분리해주었는데 expectstatus() 부분은 updateVideo에 포함해주었네요!
의도된 부분인가요?

heebong replied:
findVideo(), findVideos(), saveVideo(), updateVideo(), deleteVideo()는 테스트 함수에서 사용할 목적으로 만들어진 함수들입니다. 테스트함수는 성공과 실패들을 테스트 하기 때문에 expectstatus()까지만 포함된 함수를 만들어 테스트함수들에서 유동적으로 사용할 수 있게 했습니다! 
때문에 updateVideo()에서만 expectstatus() 부분을 추가한건 아닌데 혹시 다른 부분을 여쭤보신건가요??

gracefulBrown replied:
다른 부분도 마찬가지로 질문한거였어요ㅎㅎ
테스트 메서드를 검증할 때 expectXXX -> isXXX 세트로 검증을 하는데
status만 updateVideo(을 비롯한 테스트 메서드)로 감싸져있어서요ㅎㅎ

gracefulBrown commented on backend/src/main/java/com/woowacourse/edd/application/service/UserService.java:
```diff
@@ -0,0 +1,38 @@
+        userInternalService.delete(id);
+    }
+
+    public UserResponse findbyId(Long id) {
```
카멜 표기법 지켜주세요 :)

Conatuseus replied:
수정했습니다!

Conatuseus replied:
수정했습니다!

gracefulBrown commented on backend/src/main/java/com/woowacourse/edd/application/service/UserInternalService.java:
```diff
@@ -0,0 +1,51 @@
+
+@Service
+@Transactional
+class UserInternalService {
```
InternalService와 Service를 나누었군요
아래의 문서에 컨벤션을 정리한 것 같은데 이 부분도 같이 정리해보면 좋을 것 같아요:)

https://github.com/everyone-driven-development/miniprojects-2019/wiki/EDD-%EA%B7%9C%EC%B9%99

heebong replied:
컨벤션 추가했습니다!

gracefulBrown commented on backend/src/main/java/com/woowacourse/edd/application/service/VideoInternalService.java:
```diff
@@ -0,0 +1,51 @@
+class VideoInternalService {
+
+    private final VideoRepository videoRepository;
+    private final VideoConverter videoConverter = new VideoConverter();
```
사용하는 부분마다 생성해서 사용하나요?
static으로 사용하는건 어떨까요?

Conatuseus replied:
사용하는 부분마다 생성하는 경우보다 static으로 사용하는 것이 더 좋아 보여서
static으로 변경했습니다. 👍

gracefulBrown commented on backend/src/main/java/com/woowacourse/edd/domain/User.java:
```diff
@@ -0,0 +1,74 @@
+import org.hibernate.annotations.CreationTimestamp;
+import org.hibernate.annotations.UpdateTimestamp;
+
+import javax.persistence.*;
```
구글의 네이밍 컨벤션에는 아래와 같은 규칙이 있는데요
import에 `*`을 사용하는 것과 사용하지 않게하는 것은 어떤 차이가 있을까요?
팀에서 논의해서 규칙을 고민해보아도 좋을 것 같아요 :)

> https://google.github.io/styleguide/javaguide.html#s3.3.1-wildcard-imports

Conatuseus replied:
*을 사용하면 패키지 간의 충돌이 날 수도 있겠다는 생각을 했습니다.
두 메서드가 이름이 같을 때, 컴파일 에러가 날 수 있다고 생각해요. 그래서 변경했습니다.
👍

gracefulBrown commented on backend/src/main/java/com/woowacourse/edd/application/service/UserService.java:
```diff
@@ -0,0 +1,38 @@
+        return userConverter.toResponse(user);
+    }
+
+    public void delete(Long id) {
```
현재 클래스마다 delete가 가장 아래에 위치하는 경우와 그렇지 않은 경우가 있는데요
클래스의 메서드 순서는 어떤 기준으로 정렬하면 좋을까요?

Conatuseus replied:
같이 피드백 반영하면서 CRUD 순서로 정했습니다. 😄 💯

Conatuseus replied:
같이 피드백 반영하면서 CRUD 순서로 정했습니다. 😄 💯

gracefulBrown reviewed (changes requested):
많은 부분이 구현되고 있네요 👍 
팀 컨벤션 위주로 고민해볼 거리 위주로 코멘트를 남겼습니다.
규칙을 정해서 문서로 정리해보면 이 후 코드작업 시 고민의 시간을 조금 줄일 수 있을 것 같네요 :)
확인해보시고 궁금한 점 있으면 질문주세요 ^^

wmakerjun commented on frontend/signup.html:
```diff
@@ -0,0 +1,91 @@
+                <div class="tab-content">
+                    <div class="form-horizontal mrg-top-40 pdd-right-30">
+                        <div class="form-group row">
+                            <label for="form-1-1" class="col-md-3">닉네임</label>
```
label의 for 속성은 연결되는 form element의 id와 동일해야 합니다. form-1-1 속성은 아래에 있는 닉네임 input의 id와 통일시켜주세요 :)
https://webdir.tistory.com/319

heebong replied:
반영했습니다!

wmakerjun commented on frontend/video-channel.html:
```diff
@@ -0,0 +1,190 @@
+<script src="./js/headerTemplate.js"></script>
+<script src="./js/video/videocardtemplate.js"></script>
+<script src="./js/channelpage.js"></script>
+<script>
```
index.html에 정확히 같은 자바스크립트 코드 부분이 있어요~ 중복을 제거하고 재사용하면 좋을 것 같아요~ 😉

wmakerjun commented on frontend/video-edit.html:
```diff
@@ -157,7 +157,6 @@ <h4 class="card-title">Video 수정</h4>
     <script src="./js/video/edit.js"></script>
-    <script src="./js/video/create.js"></script>
     <script>
         addHeader(document.querySelector('.header-wrapper'));
```
index.html에 정확히 같은 자바스크립트 코드 부분이 있어요~ 중복을 제거하고 재사용하면 좋을 것 같아요~ 😉

wmakerjun commented on frontend/video-channel.html:
```diff
@@ -0,0 +1,190 @@
+
+    <link rel="stylesheet" href="./css/app.css">
+    <link rel="stylesheet" href="./css/index.css">
+    <link rel="stylesheet" href="./css/side-nav.css">
```
video-*.html 파일들에서 모두 head와 header ui 부분이 중복되고 있어요~! 중복코드들을 타임리프를 이용했던 것처럼 재사용하는 방식으로 바꿔보면 어떨까요~?

heebong replied:
head 부분 중복 제거했습니다!

kangmin46 committed:
[feat] user CRUD (#118)

* User CRUD 구현

kangmin46 committed:
[feat] add signup validation test, add exception Handler for validation annotation (#122)

* user validation 테스트코드 3개추가
* exception Handler 의 메소드 추가

Laterality committed:
[feat] add login, signup view api (#119)

* 회원가입/로그인 페이지 API연동

Laterality committed:
[fix] fix incorrect db schema (#123)

* 프로덕션 DB 스키마의 문법 오류를 수정합니다.

Laterality committed:
[fix] change db schema for production (#124)

* 프로덕션 DB 스키마의 누락된 칼럼을 추가합니다.

kangmin46 committed:
[fix] change name validation error message (#126)

* 닉네임 유효성검사 에러메시지를 수정했습니다.

Laterality committed:
[feat] correct typo (#128)

* 프로덕션 DB 스키마의 오타를 수정했습니다.

Laterality committed:
[feature] video creator mapping (#129)

* 비디오와 사용자 간 관계(동영상 게시자)를 매핑합니다.

Laterality committed:
correct typo (#130)

* 프로덕션 DB 스키마의 오타를 수정합니다.

Laterality committed:
change db schema for production (#131)

* DB 스키마가 user 테이블을 먼저 생성하도록 수정합니다.

Laterality committed:
change db schema for production (#131)

* DB 스키마가 사용자 테이블을 먼저 생성하도록 변경합니다.

kang-hyungu commented on backend/src/main/java/com/woowacourse/edd/application/converter/UserConverter.java:
```diff
@@ -0,0 +1,23 @@
+    private static final Boolean IS_DELETED = false;
+
+    public User toSaveEntity(UserRequestDto userSaveRequestDto) {
+        return new User(userSaveRequestDto.getName(), userSaveRequestDto.getEmail(), userSaveRequestDto.getPassword(), IS_DELETED);
```
사용자를 새로 생성할 때 삭제 상태가 아니다를 표시하기 위한 IS_DELETED는 비즈니스 로직이 아닐까요?
생성자에서 처리하는건 어떨까요?

heebong replied:
변경했습니다! 확인 부탁드립니다~

kang-hyungu commented on backend/src/main/java/com/woowacourse/edd/application/service/LoginInternalService.java:
```diff
@@ -0,0 +1,28 @@
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@Transactional
```
DML 처리가 없어서 아직 트랜잭션이 필요해보이진 않네요

heebong replied:
변경했습니다! 확인 부탁드립니다~

kang-hyungu commented on backend/src/main/java/com/woowacourse/edd/application/service/UserInternalService.java:
```diff
@@ -0,0 +1,60 @@
+    public void delete(Long id, Long loggedInId) {
+        checkAuthorization(id, loggedInId);
+        User user = findById(id);
+        if (user.isDeleted()) {
```
사용자 엔티티에 삭제여부 필드를 두게 되면 isDelete 같이 체크하는 로직이 생기게 되죠
하이버네이트 기능중에 where 어노테이션이 있는데 사용자 엔티티에 적용해보면 어떨까요?

heebong replied:
변경했습니다! 확인 부탁드립니다~

kang-hyungu commented on backend/src/main/java/com/woowacourse/edd/application/service/VideoService.java:
```diff
@@ -2,48 +2,51 @@
 
 @Service
-@Transactional
 public class VideoService {
```
트랜잭션이 필요하지 않을까요?

heebong replied:
변경했습니다! 확인 부탁드립니다~

kang-hyungu commented on backend/src/main/java/com/woowacourse/edd/domain/Video.java:
```diff
@@ -15,22 +16,59 @@
+    @Column(nullable = false)
     private LocalDateTime createDate;
 
+    @ManyToOne
```
지연로딩을 추가해보면 어떨까요?

heebong replied:
변경했습니다! 확인 부탁드립니다~

kang-hyungu commented on backend/src/test/java/com/woowacourse/edd/presentation/controller/VideoControllerTests.java:
```diff
@@ -108,44 +151,31 @@ private StatusAssertions findVideo(String uri) {
-        return executeGet(VIDEOS_URI + "?page="+page+"&size="+size+"&sort="+sort+","+direction)
-                .exchange()
-                .expectStatus();
+        return executeGet(VIDEOS_URI + "?page=" + page + "&size=" + size + "&sort=" + sort + "," + direction)
```
문자열로 쿼리스트링 작성하고 수정하려면 불편하죠 ㅠㅠ
자바에서 제공하는 URI 클래스를 써보면 어떨까요?

kang-hyungu reviewed (changes requested):
전체적으로 코드 구현 잘하셨어요 👏 
몇 가지 간단한 코멘트만 남겼으니 참고하세요!

Conatuseus committed:
[feedback] accommodate reviewer feedback   (#134)

- wrapper class 삭제
- parameterized test 추가
- converter을 모두 static으로
- wildcard import 방법 변경
- service 내 메서드 순서를 CRUD 순으로 정렬

Laterality committed:
Update domain and service (#135)

* 일부 서비스에서 불필요한 `@Transactional` 어노테이션을 제거합니다.
* `Video`의 연관 관계 엔티티에 지연 로딩을 적용합니다.

Conatuseus committed:
[feedback] 구구 피드백 반영 (#136)

* User Entity에 `@where` annotation 적용
* User의 생성자에서 delete flag를 초기화 하도록 변경

Laterality committed:
[fix] fix #137 (#138)

* 오타를 수정합니다.

Laterality committed:
[fix] changed as user entity constructor changed (#139)

* user entity 생성자 변경에 따른 오류 수정

gracefulBrown reviewed (approved):
피드백 반영 확인 하였습니다 👍 
고생 많으셨습니다! approve할게요~

wmakerjun commented on frontend/js/mainpage.js:
```diff
@@ -1,7 +1,7 @@
-loadVideoCards = function (filter) {
-    api.requestVideos(filter)
+loadVideoCards = function (page,size,sort) {
```
지금과 같은 방식으로 변수에 함수를 선언하는 경우 예기치 못한 에러를 발생시킬 수 있습니다. [참고] (https://www.tutorialsteacher.com/javascript/javascript-variable)

바로 함수를 선언하고 싶다면 function loadVideoCard () {} 형태로 선언해주세요 :)

heebong replied:
반영했습니다!

wmakerjun commented on frontend/js/login.js:
```diff
@@ -1,8 +1,29 @@
 const handleSignInEvent = function (event) {
```
handleSignInEvent 함수에서 event 객체를 사용하지 않는다면 지워주세요 :)

heebong replied:
반영했습니다!

wmakerjun commented on frontend/js/video/detail.js:
```diff
@@ -16,4 +16,23 @@ const detailVideo = function(data) {
+detailView();
+
+document.querySelector('#btn-edit').addEventListener('click', function(event) {
+    location.href = '/video-edit.html?id=' + wootubeCtx.util.getUrlParams().id
```
url응 [문자열 리터럴](https://poiemaweb.com/es6-template-literals)을 참고해서 표현하면 좋을 것 같아요!

heebong replied:
반영했습니다!

wmakerjun commented on frontend/js/channelpage.js:
```diff
@@ -0,0 +1,45 @@
+    }
+}
+
+loadVideoCards = function (page, size, sort) {
```
지금과 같은 방식으로 변수에 함수를 선언하는 경우 예기치 못한 에러를 발생시킬 수 있습니다. [참고] (https://www.tutorialsteacher.com/javascript/javascript-variable)

바로 함수를 선언하고 싶다면 function loadVideoCards () {} 형태로 선언해주세요 :)

heebong replied:
변경했습니다!

wmakerjun commented on frontend/video-edit.html:
```diff
@@ -157,7 +157,6 @@ <h4 class="card-title">Video 수정</h4>
     <script src="./js/common.js"></script>
```
https://github.com/woowacourse/miniprojects-2019/blob/cd183e91a1a60833e5c9acd33353e837fd31aa54/frontend/video-edit.html#L121

html 코드 쪽에 label의 for 속성은 연결되는 form element의 id와 동일해야 합니다. form-1-1 속성은 아래에 있는 input tag의 id와 통일시켜주세요 :)

heebong replied:
반영했습니다!

wmakerjun reviewed (changes requested):
늦은 시간까지 프로젝트 진행하느라 빠븐 가운데 
많은 양의 프론트엔드 코드를 높은 품질로 구현하느라 너무너무너무 고생하셨습니다 👏 👍 💯 
더 좋은 마무리를 위해서 몇가지 반복해서 실수할 수 있는 부분들에 대한 피드백을 남겼습니다 :)

1. 변수 선언방식
2. 문자열 리터럴
3. 안쓰는 파라미터 삭제 (특히 event 객체)
4. 중복 html 코드 제거
5. 중복 script 제거
6. form에서 label의 for 속성 이해

위 6가지를 다시한번 체크하면서 피드백을 반영해주시길 바래요 :) 늘 응원합니다 EDD조!!!! 화이팅!!

Laterality committed:
[fix] fix issue (#144)

* 개발 환경의 CORS 이슈를 해결합니다

kang-hyungu reviewed (approved):
피드백 반영 잘 하셨어요!! 👍

Laterality committed:
[feedback] feedback gugu

- 쿼리 문자열을 포함하는 URI의 생성 방식을 변경합니다

heebong committed:
[feedback] feedback jun (#146)

* 변수 선언방식 const로 변경
* 문자열 리터럴 사용
* 안쓰는 event 파라미터 삭제
* 중복 html 코드 head 제거
* 중복 script sideNav 제거
* form에서 label의 for와 input의 id 값 동일하게 변경
